gnucash/gnome/gnc-plugin-page-account-tree.c

```diff
@@ -1499,167 +1670,64 @@ gnc_plugin_page_account_tree_cmd_delete_account (GtkAction *action, GncPluginPag
+                error_dialog = gtk_message_dialog_new (GTK_WINDOW(window),
+                                                       GTK_DIALOG_DESTROY_WITH_PARENT,
+                                                       GTK_MESSAGE_ERROR,
+                                                       GTK_BUTTONS_NONE,
+                                                       "%s", message);
+                gtk_dialog_add_buttons (GTK_DIALOG(error_dialog),
+                                        _("_Pick another account"), GTK_RESPONSE_CANCEL,
+                                        _("_Do it anyway"), GTK_RESPONSE_ACCEPT,
+                                        (gchar *)NULL);
+                response = gtk_dialog_run (GTK_DIALOG (error_dialog));
+                gtk_widget_destroy (error_dialog);
+                g_free (message);
+                if (response == GTK_RESPONSE_ACCEPT) break;
             }
-            else if (NULL != sta)
-            {
-                /* Move the splits of its subaccounts, if any. */
-                gnc_account_foreach_descendant(account,
-                                               (AccountCb)xaccAccountMoveAllSplits,
-                                               sta);
-            }
-            if (NULL != ta)
-            {
-                /* Move the splits of the account to be deleted. */
-                xaccAccountMoveAllSplits (account, ta);
-            }
-            xaccAccountCommitEdit (account);
-
-            /* Drop all references from the state file for
-             * any subaccount the account still has
-             */
-            acct_list = gnc_account_get_children(account);
-            for (ptr = acct_list; ptr; ptr = g_list_next(ptr))
-            {
-                guid = xaccAccountGetGUID (ptr->data);
-                guid_to_string_buff (guid, guidstr);
-                gnc_state_drop_sections_for (guidstr);
-            }
-            g_list_free(acct_list);
-
-            /* Drop all references from the state file for this account
-             */
-            guid = xaccAccountGetGUID (account);
-            guid_to_string_buff (guid, guidstr);
-            gnc_state_drop_sections_for (guidstr);
-
-            /*
-             * Finally, delete the account, any subaccounts it may still
-             * have, and any splits it or its subaccounts may still have.
-             */
-            xaccAccountBeginEdit (account);
-            xaccAccountDestroy (account);
-            gnc_resume_gui_refresh ();
-            gnc_unset_busy_cursor(NULL);
         }
-    }
-    g_free(acct_name);
+        gtk_widget_destroy(dialog);
+        g_list_free(filter);
+    } /* (NULL != splits) || (NULL != children) */
+    gnc_plugin_page_account_tree_cmd_delete_account_int (action, page, ta, sta, saa, acct_name, delete_res);
 }
```

Comment:
Requiring functions to sink their parameters is a good way to have leaks or double-frees because it's too hard to keep track of whether they do or not. Pass `acct_name` as `const char*` and free it when the function returns... but see my note about allocating it where it's used since you're retrieving the account from the page there anyway. That way you could free it when it's last used in this function.

Aside from having an unwieldy name, `gnc_plugin_page_account_tree_cmd_delete_account_int` does too many things. A better decomposition would have
```
if (confirm_account_delete_dialog (page, ta, sta, saa, delete_res.has_splits))
{
    if (ta || sta || saa)
       move_subaccounts (account, ta, sta, saa);
    delete_account (account);
}
```